pkg/composition/controller2.go

```diff
@@ -0,0 +1,624 @@
+		_, retriableStatus, errStatus := c.updateServiceDescriptor(logger, sd)
+		if errStatus != nil {
+			return retriableStatus, errStatus
+		}
+	}
+
+	// First filter to filter out the service descriptor if namespace is specified, to avoid
+	// treading on compositions trying to handleProcessResult and overwriting each other.
+	if c.namespace != "" {
+		nsServiceName, _ := deconstructNamespaceName(c.namespace)
+		if serviceName != nsServiceName {
+			logger.Sugar().Infof("Namespace %q can't possibly be created by SD %q. Skipping.", c.namespace, sd.Name)
+			return false, nil
+		}
+	}
+
+	filteredTransformer := &FilteredTransformer{
+		Transformer: c.sdTransformer,
+		Remove: func(info FormationObjectInfo) bool {
+			formLogger := logger.With(logz.NamespaceName(info.Namespace))
+			formLogger.Debug("Processing LocationDescriptor for ServiceDescriptor")
+			// filter the formationObjectDefList so we only handle the namespace we know
+			if info.Namespace != c.namespace && c.namespace != "" {
+				formLogger.Sugar().Infof("Namespace %q doesn't match namespace %q observed by the controller. Skipping formationObjectDef %v.",
+					info.Namespace, c.namespace, info.Location.Label)
+				return true
+			}
+			return false
+		},
+	}
+
+	formationObjectDefList, err := ProcessSD(sd, filteredTransformer)
+
+	foResults := make([]formationObjectResult, 0, len(formationObjectDefList))
+
+	if err != nil {
+		conflict, processRetryable, processErr := c.handleProcessResult(logger, serviceName, sd, foResults, false, false, err)
+		if conflict {
+			return false, nil
+		}
+
+		return processRetryable, processErr
+	}
+
+	var (
+		results        []*formationObjectResult
+		retryable      = false
+		deleteFinished = true
+		processFailure error
+	)
+
+	for _, result := range foResults {
+		formLogger := logger.With(logz.NamespaceName(result.namespace.Namespace))
+		formLogger.Debug("Processing LocationDescriptor for ServiceDescriptor")
+
+		var (
+			newResult        *formationObjectResult
+			processConflict  bool
+			processRetryable bool
+			processErr       error
+		)
+		if sd.ObjectMeta.DeletionTimestamp != nil {
+			var itemFinished bool
+			itemFinished, processConflict, processRetryable, newResult, processErr = c.delete(logger, sd, &result)
+			deleteFinished = deleteFinished && itemFinished
+		} else {
+			processConflict, processRetryable, newResult, processErr = c.processNormalFormationObjectDef2(formLogger, sd, &result)
+		}
+
+		if processConflict {
+			// Something ran into a conflict during processing
+			// The entire ServiceDescriptor will be processed anyway so we just exit
+			return false, nil
+		}
+
+		if processErr != nil {
+			// stop the presses
+			deleteFinished = false
+			processFailure = processErr
+			retryable = processRetryable
+			break
+		}
+		results = append(results, newResult)
+	}
+
+	// If deletion is complete, remove finalizer
+	if sd.ObjectMeta.DeletionTimestamp != nil && deleteFinished {
```

Comment:
Move the finalizer logic out of status calculation since it doesn't really have anything to do with updating the status and has more to do with patching over the delete story.